src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/ReviewUnusedParameters.cs

```diff
@@ -70,62 +64,111 @@ public override void Initialize(AnalysisContext context)
-                           eventsArgSymbol, methodsUsedAsDelegates, attributeSetForMethodsToIgnore);
-                    }
-                });
+        private static void AddParameters(IMethodSymbol method, ConcurrentDictionary<IParameterSymbol, bool> unusedParameters)
+        {
+            foreach (var parameter in method.Parameters)
+            {
+                unusedParameters.TryAdd(parameter, false);
+            }
+        }
 
-                // Register a compilation end action to filter all methods used as delegates and report any diagnostics
-                compilationStartContext.RegisterCompilationEndAction(compilationAnalysisContext =>
+        private static void ReportUnusedParameters(
+            SymbolAnalysisContext symbolEndContext,
+            ConcurrentDictionary<IParameterSymbol, bool> parameterUsageMap,
+            ConcurrentDictionary<IMethodSymbol, bool> methodsUsedAsDelegates)
+        {
+            // Report diagnostics for unused parameters.
+            foreach (var (parameter, used) in parameterUsageMap)
+            {
+                if (used)
                 {
-                    // Report diagnostics for unused parameters.
-                    var unusedParameters = unusedMethodParameters.Where(kvp => !methodsUsedAsDelegates.Contains(kvp.Key)).SelectMany(kvp => kvp.Value);
-                    foreach (var parameter in unusedParameters)
-                    {
-                        var diagnostic = Diagnostic.Create(Rule, parameter.Locations[0], parameter.Name, parameter.ContainingSymbol.Name);
-                        compilationAnalysisContext.ReportDiagnostic(diagnostic);
-                    }
+                    continue;
+                }
 
-                });
-            });
+                var containingMethod = (IMethodSymbol)parameter.ContainingSymbol;
+
+                // Don't report parameters for methods used as delegates.
+                // We assume these methods have signature requirements,
+                // and hence its unused parameters cannot be removed.
+                if (methodsUsedAsDelegates.ContainsKey(containingMethod))
+                {
+                    continue;
+                }
+
+                // Do not flag unused 'this' parameter of an extension method.
+                if (containingMethod.IsExtensionMethod && parameter.Ordinal == 0)
+                {
+                    continue;
+                }
+
+                // Do not flag unused parameters with special discard symbol name.
+                if (parameter.IsSymbolWithSpecialDiscardName())
+                {
+                    continue;
+                }
+
+                var diagnostic = parameter.CreateDiagnostic(Rule, parameter.Name, parameter.ContainingSymbol.Name);
+                symbolEndContext.ReportDiagnostic(diagnostic);
+            }
         }
 
-        private static void AnalyzeMethod(
+#pragma warning disable RS1012 // Start action has no registered actions.
```

Comment:
Too bad there is no interface for the various context so that we can avoid the issue here.